### PR TITLE
[CORE] Add custom cost evaluator for optimize buildSide of shuffled hash join

### PR DESCRIPTION
**Pull request by zml1206**
## What changes were proposed in this pull request?

Split from #6093 .
Add custom cost evaluator for use the smaller table to build hashmap in shuffled hash join.

### CODE REVIEW

**Comment by zml1206** [comment]
cc @zhztheplayer @ulysses-you thank you.

File: shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala [hunk 1]
```diff
@@ -1962,6 +1966,17 @@ object GlutenConfig {
...
       .createWithDefault(true)
 
+  val COST_EVALUATOR_ENABLED =
+    buildConf(GlutenConfig.GLUTEN_COST_EVALUATOR_ENABLED)
+      .internal()
+      .doc(
+        "If true and gluten enabled, use " +
+          "org.apache.spark.sql.execution.adaptive.GlutenCostEvaluator as custom cost " +
+          "evaluator class, else follow the configuration " +
+          "org.apache.spark.sql.execution.adaptive.GlutenCostEvaluator.")
```
Comment (zhztheplayer) [inline shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala:1976]: `org.apache.spark.sql.execution.adaptive.GlutenCostEvaluator` this should be corrected

Comment (zml1206) [reply shims/common/src/main/scala/org/apache/gluten/GlutenConfig.scala]: My mistake, fixed, thank you.

**Review by zhztheplayer — APPROVED** [review]
Thanks!

File: shims/spark35/src/main/scala/org/apache/spark/sql/execution/adaptive/GlutenCostEvaluator.scala [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+import org.apache.spark.sql.catalyst.SQLConfHelper
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.internal.SQLConf
+
+/** This [[CostEvaluator]] is to force use the new physical plan when cost is equal. */
+case class GlutenCostEvaluator() extends CostEvaluator with SQLConfHelper {
+  override def evaluateCost(plan: SparkPlan): Cost = {
+    val forceOptimizeSkewedJoin = conf.getConf(SQLConf.ADAPTIVE_FORCE_OPTIMIZE_SKEWED_JOIN)
+    new GlutenCost(SimpleCostEvaluator(forceOptimizeSkewedJoin), plan)
```
Comment (ulysses-you) [inline shims/spark35/src/main/scala/org/apache/spark/sql/execution/adaptive/GlutenCostEvaluator.scala:27]:
```scala
if (GlutenConfig.getConf.enableGluten) {
  GlutenCost(SimpleCostEvaluator(forceOptimizeSkewedJoin), plan)
} else {
  SimpleCostEvaluator(forceOptimizeSkewedJoin)
}

```

To make it as a query level config.

**Review by ulysses-you — APPROVED** [review]
lgtm if tests pass

**Comment by zml1206** [comment]
Thank you for review. @ulysses-you @zhztheplayer

**Pull request merged**